File: panpiper/workflow/scripts/checkm_bin-stats.R
# ----------------------------------------------------------------------------
# Copyright (c) 2022--, Panpiper development team.
#
# Distributed under the terms of the Modified BSD License.
#
# The full license is in the file COPYING.txt, distributed with this software.
# ----------------------------------------------------------------------------
# Title: checkm_bin-stats.R- program to load checkM results from one file and graph
# Author- Renee Oles
# Purpose: R Script to read the CheckM results: bin_stats.analyze.tsv and concatenated.tre

require(tidyverse)

args <- commandArgs(trailingOnly = TRUE)
input <- args
out_pref <- gsub("Quality.*", "", args[1])
output <- paste(out_pref,"Quality/CheckM/checkm_stats",sep="")

checkm <- data.frame()
for (i in input){
  line <- read.delim(i, header=FALSE)
  i <- gsub("../Assembly/Shovill/","",i)
  i <- gsub("/storage/bin_stats.analyze.tsv","",i)
  line$Sample <- i  
  checkm <- rbind(checkm,line)
}
checkm2 <- str_split_fixed(checkm$V2, ",", 15)
checkm2<-apply(checkm2,2,function(x) gsub("'","",as.character(x)))
checkm2<-apply(checkm2,2,function(x) gsub("\\{","",as.character(x)))
checkm2<-apply(checkm2,2,function(x) gsub("\\}","",as.character(x)))
checkm2<-apply(checkm2,2,function(x) sub(".*\\:","",as.character(x)))
checkm2 <- as.data.frame(checkm2)
checkm <- cbind(checkm[,3],checkm2)
colnames(checkm) <- c("Sample", "GC", "GC std", "Genome size",  "ambiguous bases",  "scaffolds",  "contigs",  "Longest scaffold",
                      "Longest contig", "N50 scaffolds",  "N50 contigs",  "Mean scaffold length", "Mean contig length",
                      "Coding density", "Translation table",  "predicted genes")
checkm$Sample <- basename(checkm$Sample)
write.table(checkm,paste(output,".txt",sep=""),quote=FALSE,col.names = TRUE,row.names = FALSE, sep="\t")

File: panpiper/workflow/scripts/checkm_bin-stats.py
import sys
import os
import pandas as pd

def process_checkm_files(output_prefix: str, input_files: list):
    """
    Process CheckM output files and generate a consolidated statistics file.

    Args:
        output_prefix (str): Prefix for the output file path.
        input_files (list): List of input file paths.

    Returns:
        None
    """
    # Define output path
    output_path = os.path.join(output_prefix, "Quality/CheckM/checkm_stats.txt")

    checkm = pd.DataFrame()

    # Read and process input files
    for input_file in input_files:
        line = pd.read_csv(input_file, sep='\t', header=None)
        checkm = pd.concat([checkm, line], ignore_index=True)

    # Process the checkm2 column
    checkm2 = checkm.drop(checkm.columns[0], axis=1)
    checkm2 = checkm2.astype(str).applymap(lambda x: str(x).replace("'", "").replace("{", "").replace("}", ""))
    checkm2 = checkm2[1].str.split(',', expand=True)
    checkm2 = checkm2.applymap(lambda x: x.split(":")[1])
    checkm2.columns = ["GC", "GC std", "Genome size", "ambiguous bases", "scaffolds", "contigs",
                       "Longest scaffold", "Longest contig", "N50 scaffolds", "N50 contigs", "Mean scaffold length",
                       "Mean contig length", "Coding density", "Translation table", "predicted genes"]

    # Process the Sample column
    checkm2.insert(0, "Sample", checkm[0])

    # Save the DataFrame to a tab-separated text file
    checkm2.to_csv(output_path, sep='\t', index=False)

# Get input file paths from command-line arguments
input_files = sys.argv[2:]

# Define output path
output_prefix = sys.argv[1]

# Call the function
process_checkm_files(output_prefix, input_files)
